src/Symfony/Component/Security/Http/Authenticator/RemoteUserAuthenticator.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Authenticator;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+/**
+ * This authenticator authenticates a remote user.
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Maxime Douailin <maxime.douailin@gmail.com>
+ *
+ * @final
+ *
+ * @internal in Symfony 5.1
+ */
+class RemoteUserAuthenticator extends AbstractPreAuthenticatedAuthenticator
+{
+    private $userKey;
+
+    public function __construct(UserProviderInterface $userProvider, string $userKey = 'REMOTE_USER', ?LoggerInterface $logger = null)
+    {
+        parent::__construct($userProvider, $logger);
+
+        $this->userKey = $userKey;
+    }
+
+    protected function extractUsername(Request $request): ?string
+    {
+        return $request->server->get($this->userKey);
```

Comment:
As I mentioned on the base class, in the old system, both this and X509 would throw a `BadCredentialsException` if this didn't exist. This seems a little odd to me, but apparently that's how it's supposed to work.